### DISCUSSION TITLE
Permit GeoSearch to rank results based on type

### DISCUSSION THREAD

**Discussion by milespetrov** [fgpv-vpgf/fgpv-vpgf#4114] [Team Posts]
Original issue: https://github.com/fgpv-vpgf/fgpv-vpgf/issues/3307

## The problem

A map author wants geosearch results to be filtered and sorted in a way which suits the theme of the map data - i.e. a map showcasing downstream water flows would emphasize River, Lake, and Spring results over other types, and omit some results like Air Navigation Features altogether. 

## Solution summary 

We can allow the map author, through the ramp config, to set certain parameters for the Canadian GeoNames Search Service API like `category`, `theme`, `concise`, and `sort-field`. These config properties would be passed through to the geosearch library.

```json
{
  "geoSearch": {
    "category": "O",
    "concise": ["CITY", "AIR", "BAY"],
    "sort-field": "name"
  }
}
```

See: https://www.nrcan.gc.ca/earth-sciences/geography/place-names/tools-applications/9249

## Enhancing the "concise" config property

The Canadian GeoNames Search Service API in conjunction with `sort-field=concise` sorts results based on the feature type name (i.e. `AIR -> BAY -> CITY`) but this sort order cannot be configured. Therefore ramp would need to sort the results manually, based on the order provided in the ramp config.

Further enhancement can include an `excludeConcise` type config property which includes all feature types not in the list provided.

See: http://geogratis.gc.ca/services/geoname/en/codes/concise

## Handling more than 1k results

The Canadian GeoNames Search Service API returns up to 1,000 results per request. Additional results would require additional requests with the `skip` parameter set. If we perform sorting within ramp, we would need to implement this functionality so we have the entire data set to sort. 

### Better/worse options?

- Make a separate request for each `concise` item then sort locally
- Don't bother fetching additional results, just sort what you have

## Relevance of "theme"

See: http://geogratis.gc.ca/services/geoname/en/codes/theme

The list of supported themes doesn't seem exhaustive, and perhaps should be omitted from the possible geosearch parameters in the ramp config. This would leave `concise` as the sole way to both filter and sort results.

## Location of local sorting

Local sorting can happen either in the geosearch library, or within ramp. It seems best to keep all this logic within geosearch and ramp blindly renders the filtered/sorted results.

## To be decided:

1. Do we like the approach of adding `category`, `concise`, and `sort-field` to the ramp config?
2. Should the `concise` ramp config list dictate the sort order if `sort-field` is omitted or is set to `sort-field=concise`? 
3. Do we handle more than 1k results?
4. Is it worth keeping the `theme` config property?
5. Where should the sorting logic take place?

**Comment by james-rae** [comment]
- A 1k limit seems reasonable. I'd imagine only the most general searches break it (like "Lake").
- Separate request for each concise seems overkill; if anything, this should be a flag to turn on for very rare scenarios
- For lists of concise codes in config file, lets pick a more general term / meaningful term (e.g. subcategory). We can have different search libraries; let our geogratis library map subcategory to "concise". If Jimmy McRampauthor is setting up his first config, I can see him going "concise what?".
- In general, anything put in the config file should have a flexible feel, and the schema description should make it clear that support is implementation specific.
- Theme seems interesting but maybe as a stretch goal / next version consideration? E.g. see how well the concise code sorting holds up in a live app that uses it.
- Sorting logic inside library seems correct.
- We mentioned ability to omit certain concise codes but then never came back to it. Another stretch-goal?

That said, the whole "general search library" connected with specific UI elements and config schema in the viewer seems to have limitations.  E.g. wanting to use a different search service that has an extra parameter would mean updates to ramp. Might be going too far down the rabbit hole of generic programming and we just accept that we only support search services with certain capabilities (probably fine).  But if we agree on that, then maybe some of the suggestions above change (e.g. may as well always sort in the client instead of implementing it in every library).

**Comment by mweech** [comment]
My thoughts:
1. I think we should support category, concise and sort-field arguments, but as James points out, using friendlier/more appropriate names (concise and category are both odd definitions and might have been lost in translation from French)
2. It's clever, but not very intuitive.  I think we should have a config value called "categories" that lists the categories that would be included, and have an optional value for "sort order".  "sort-field" would need to remain as a seperate argument that is specific to Geogratis, while "categories" would affect post-geogratis query sorting.
3. Agree with James, 1K search limit is enough
4. Yes, but we can add this later.  
5. Sorting at Geosearch Library level makes sense to me..

**Comment by mweech** [comment]
### Acceptance Criteria
1. RAMP configuration options for Geosearch will be added to support both the definition of a limited subset of categories, as well as defining an initial prioritization of search result categories following:
    - "categories" (not required) which will define the arguments that map to the "concise" parameter of geogratis: http://geogratis.gc.ca/services/geoname/en/codes/concise
    - "sort-order" (not required) which will define the sort order of defined "categories". If a category is missing from the "sort-order", it will be appended to the bottom of the sorted list.  A "sort-order" can be defined without limiting results via "categories"
    - "official-only" will be a boolean that would map to the "category" parameter of geogratis
2. Geosearch results will be limited to 1K results based on geogratis service limitations
3. "Theme" parameter will not be supported at this time.
4. Sorting of results will take place while processing results from Geogratis service and be considered tied to the service being queried. A generic search results sorting function is not required.
5. Help documentation has been updated
6. RAMP Documentation has been updated to reflect changes
